base/libgit2/libgit2.jl

```diff
@@ -386,51 +386,91 @@ function merge!(repo::GitRepo;
+                throw(GitError(Error.Merge, Error.ERROR,
+                               "Repository HEAD is detached. Remote tracking branch cannot be used."))
+            end
+            if isorphan(repo)
+                # this isn't really a merge, but really moving HEAD
+                # https://github.com/libgit2/libgit2/issues/2135#issuecomment-35997764
+                # try to figure out remote tracking of orphan head
+
+                m = with(GitReference(repo, Consts.HEAD_FILE)) do head_sym_ref
+                    match(r"refs/heads/(.*)", fullname(head_sym_ref))
                 end
-            else # try to get tracking remote branch for the head
-                if !isattached(repo)
+                if m === nothing
                     throw(GitError(Error.Merge, Error.ERROR,
-                                   "Repository HEAD is detached. Remote tracking branch cannot be used."))
+                                   "Unable to determine name of orphan branch."))
+                end
+                branchname = m.captures[1]
+                remotename = with(GitConfig, repo) do cfg
+                    LibGit2.get(String, cfg, "branch.$branchname.remote")
+                end
+                obj = with(GitReference(repo, "refs/remotes/$remotename/$branchname")) do ref
+                    LibGit2.Oid(ref)
+                end
+                with(get(GitCommit, repo, obj)) do cmt
+                    LibGit2.create_branch(repo, branchname, cmt)
                 end
-                with(upstream(head_ref)) do tr_brn_ref
-                    if tr_brn_ref === nothing
-                        throw(GitError(Error.Merge, Error.ERROR,
-                                       "There is no tracking information for the current branch."))
+                return true
+            else
+                with(head(repo)) do head_ref
+                    with(upstream(head_ref)) do tr_brn_ref
+                        if tr_brn_ref === nothing
+                            throw(GitError(Error.Merge, Error.ERROR,
+                                           "There is no tracking information for the current branch."))
+                        end
+                        [GitAnnotated(repo, tr_brn_ref)]
                     end
-                    [GitAnnotated(repo, tr_brn_ref)]
                 end
             end
         end
+    end
 
-        try
-            merge!(repo, upst_anns, fastforward,
-                   merge_opts=merge_opts,
-                   checkout_opts=checkout_opts)
-        finally
-            map(finalize, upst_anns)
-        end
+    try
+        merge!(repo, upst_anns, fastforward,
+               merge_opts=merge_opts,
+               checkout_opts=checkout_opts)
+    finally
+        map(finalize, upst_anns)
     end
 end
 
-""" git rebase --merge [--onto <newbase>] [<upstream>] """
-function rebase!(repo::GitRepo, upstream::AbstractString="", newbase::AbstractString="")
```

Comment:
guess we should add that to the list in #19839 since it would be a useful feature to have at some point